### PR TITLE
Combining Multiple Tables 

### LINKED ISSUE #25: Write tutorial for Combining data from multiple tables
## Helper information

**Topic** Combining data from multiple tables

**Starter markdown file**

https://github.com/ploomber/sql/blob/main/course-material/intro-to-sql/combining-data-from-multiple-tables.md

This notebook is blocked by #23 

## Task:

Write a tutorial that includes an exercise with hints. Assume the tutorial is written for someone who would launch the notebook on Google Colab or Binder, and would interact with the tutorial via the Jupyter notebook. Ensure to include an introduction to the problem explored. Link resources you think are relevant.

### PR DESCRIPTION
**Pull request by TegveerG**
## Describe your changes

Added tutorial and exercise questions - not fully complete yet

## Issue number

Closes #25 

## Checklist before requesting a review

- [ ] Performed a self-review of my code
- [ ] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)

----
:books: Documentation preview :books:: https://ploomber-sql--27.org.readthedocs.build/en/27/

### CODE REVIEW

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 1]
```diff
@@ -1,14 +1,305 @@
...
+
+```{code-cell} ipython3
+import csv
+import urllib.request
+import zipfile
+import os
+
+
+def extract_asc_to_csv(url, output_folder):
+    # Download the ZIP file
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:54]: I gave @jpjon  feedback on how to better document the function,and a suggestion to have this function as a stand alone script 

If we need to make fixes to the function,we run into issues when having to fix it in multiple places.

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 2]
```diff
@@ -1,14 +1,305 @@
...
+
+            print(f"Converted {asc_path} to CSV.")
+
+    print("All ASC files converted to CSV.")
+
+
+# Example usage
+link = "http://sorry.vse.cz/~berka/challenge/pkdd1999/data_berka.zip"
+output = "expanded_data"
+extract_asc_to_csv(link, output)
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:84]: It's not clear whether expanded_data is a relative path. 

Do you have this in your local files?

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 3]
```diff
@@ -1,14 +1,305 @@
...
+```
+
+After running this code, you should have the folder `expanded_data` in the current directory. Within this folder, shall be present the eight converted `.csv` files for our analysis. However, we shall focus on a subset of these eight files.
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:87]: Ah, the code is generating a folder called "expanded_data"?

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 4]
```diff
@@ -1,14 +1,305 @@
...
+
+## Load Engine
+
+We now load in our SQL extension that allows us to execute SQL queries in Jupyter Notebooks. 
+
+```{tip}
+<b>Note</b> Ensure you restart any previous notebook that has the same database name as the one initialized below.
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:94]: Why not give it a different name?

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 5]
```diff
@@ -1,14 +1,305 @@
...
+
+<!-- Now that we have renamed the columns of the `district` table, we can now combine the `account`, `card`, `district`, and `link` tables. We can do this by using the `CREATE VIEW` syntax in DuckDB. The `CREATE VIEW` syntax allows us to create a virtual table that is a result of a query. 
+
+```{important}
+The `CREATE VIEW` syntax is not supported by all databases. Be sure to look at the documentation of the database you are using to see if it supports this feature.
+```
+
+```{code-cell} ipython3
+%%sql   
+CREATE VIEW s1.combined_tables AS
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:183]: The inner join was covered in @jpjon's tutorial. 

Can you suggest other ways to combine data?

Maybe we can focus the inner, outer and other joins in your tutorial to allow for a more comprehensive set of examples?

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 6]
```diff
@@ -1,14 +1,305 @@
...
+
+The `FULL OUTER JOIN` will return all rows from both tables, and if there are no matches, the columns from the other table will be filled with `NULL` values. This is in contrast to the `INNER JOIN`, which only returns rows that have matching values in both tables and, hence, no `NULL` values.
+
+Code to implement the `FULL OUTER JOIN` is shown below:
+
+```{code-cell} ipython3
+%%sql
+SELECT a.account_id, c.card_id, d.district_id, l.disp_id
+FROM s1.link AS l
+FULL OUTER JOIN s1.card AS c
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:261]: I feel that questioning people about the full outer join is a bit unfair if we haven't shown them examples.

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md [hunk 7]
```diff
@@ -1,14 +1,305 @@
...
+
+<b>Hint</b> `COUNT()` is a aggregating function in SQL (more on aggregation later!). Try experimenting with `COUNT()` in your `SELECT` clause to see if you can find the correct count.
+
+<!-- #region -->
+<details>
+<summary>Answers</summary>
+
+```{code-cell} ipython3
+%%sql
+SELECT
```
Comment (lfunderburk) [inline course-material/intro-to-sql/combining-data-from-multiple-tables.md:285]: This suggestion feels disjointed from the hint and the question.

Select won't create a table

**Review by lfunderburk — CHANGES_REQUESTED** [review]
This is an awesome start! I love  seeing the synergy and colab. 

I can also see there is overlap between this notebook and the advanced aggregation functions.

Discuss with @jpjon if he's open to leave joins for this section instead as his section is quite comprehensive. 

Main feedback for Jonathan was packaging data download functions so that they are more easily reusable.

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md (hunk 3 above)
Comment (TegveerG) [reply course-material/intro-to-sql/combining-data-from-multiple-tables.md:87]: Yep

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md (hunk 5 above)
Comment (TegveerG) [reply course-material/intro-to-sql/combining-data-from-multiple-tables.md:183]: ok sure, can focus more on natural and outer joins

File: course-material/intro-to-sql/combining-data-from-multiple-tables.md (hunk 6 above)
Comment (TegveerG) [reply course-material/intro-to-sql/combining-data-from-multiple-tables.md:261]: shall include clear examples for it

**Pull request closed without merging**